pkg/codegen/go/gen.go

```diff
@@ -2121,6 +1989,170 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource, generateReso
 	return nil
 }
 
+func (pkg *pkgContext) genMethod(resourceName string, method *schema.Method, w io.Writer) {
+	name := resourceName
+	methodName := Title(method.Name)
+	f := method.Function
+
+	// Either returnPlainType is set, or objectReturnType, not both.
+	returnPlainType, doReturnPlainType := f.ReturnsPlainType()
+	var objectReturnType *schema.ObjectType
+	if f.ReturnType != nil {
+		if objectType, ok := f.ReturnType.(*schema.ObjectType); ok && objectType != nil {
+			objectReturnType = objectType
+		}
+	}
+
+	liftReturn := pkg.liftSingleValueMethodReturns && objectReturnType != nil && len(objectReturnType.Properties) == 1
+
+	var args []*schema.Property
+	if f.Inputs != nil {
+		for _, arg := range f.Inputs.InputShape.Properties {
+			if arg.Name == "__self__" {
+				continue
+			}
+			args = append(args, arg)
+		}
+	}
+
+	// Now emit the method signature.
+	argsig := "ctx *pulumi.Context"
+	if len(args) > 0 {
+		argsig = fmt.Sprintf("%s, args *%s%sArgs", argsig, name, methodName)
+	}
+	var retty string
+	if doReturnPlainType {
+		t := pkg.typeString(codegen.ResolvedType(returnPlainType))
+		retty = fmt.Sprintf("(o %s, e error)", t)
+	} else if objectReturnType == nil {
+		retty = "error"
+	} else if liftReturn {
+		retty = fmt.Sprintf("(%s, error)", pkg.outputType(objectReturnType.Properties[0].Type))
+	} else {
+		retty = fmt.Sprintf("(%s%sResultOutput, error)", name, methodName)
+	}
+	fmt.Fprintf(w, "\n")
+	printCommentWithDeprecationMessage(w, f.Comment, f.DeprecationMessage, false)
+	fmt.Fprintf(w, "func (r *%s) %s(%s) %s {\n", name, methodName, argsig, retty)
+
+	resultVar := "_"
+	if objectReturnType != nil {
+		resultVar = "out"
+	}
+
+	// Make a map of inputs to pass to the runtime function.
+	inputsVar := "nil"
+	if len(args) > 0 {
+		inputsVar = "args"
+	}
+
+	// Now simply invoke the runtime function with the arguments.
+	outputsType := "pulumi.AnyOutput"
+	if objectReturnType != nil || doReturnPlainType {
+		if liftReturn {
+			outputsType = fmt.Sprintf("%s%sResultOutput", cgstrings.Camel(name), methodName)
+		} else {
+			outputsType = fmt.Sprintf("%s%sResultOutput", name, methodName)
+		}
+	}
+
+	if !doReturnPlainType {
+		fmt.Fprintf(w, "\t%s, err := ctx.Call(%q, %s, %s{}, r)\n", resultVar, f.Token, inputsVar, outputsType)
+	}
+
+	if doReturnPlainType {
+		fmt.Fprintf(w, "\tctx.CallReturnPlainResource(%q, %s, %s{}, r, reflect.ValueOf(&o), &e)\n",
```

Comment:
CallReturnPlainResource is no longer a good name since we're returning any values not just a Resource. Also consider moving this out of SDK and into codegen utilities module.